src/sentry/spans/grouping/strategy/base.py

```diff
@@ -0,0 +1,182 @@
+    from mypy_extensions import TypedDict
+else:
+
+    def TypedDict(*args, **kwargs):
+        pass
+
+
+Span = TypedDict(
+    "Span",
+    {
+        "trace_id": str,
+        "parent_span_id": str,
+        "span_id": str,
+        "start_timestamp": float,
+        "timestamp": float,
+        "same_process_as_parent": bool,
+        "op": str,
+        "description": Optional[str],
+        "fingerprint": Optional[Sequence[str]],
+        "tags": Optional[Any],
+        "data": Optional[Any],
+    },
+)
+
+
+# A callable strategy is a callable that when given a span, it tries to
+# returns a fingerprint. If the strategy does not apply to the span, it
+# should return `None` to indicate that the strategy should not be used
+# and to try a different strategy. If the strategy does apply, it should
+# return a list of strings that will serve as the span fingerprint.
+CallableStrategy = Callable[[Span], Optional[Sequence[str]]]
+
+
+@dataclass(frozen=True)
+class SpanGroupingStrategy:
+    name: str
+    # The strategies to use with the default fingerprint
+    strategies: Sequence[CallableStrategy]
+
+    def execute(self, spans: Sequence[Span]) -> Dict[str, str]:
+        return {span["span_id"]: self.get_span_group(span) for span in spans}
+
+    def get_span_group(self, span: Span) -> str:
+        fingerprints = span.get("fingerprint") or ["{{ default }}"]
+
+        result: List[str] = []
+
+        for fingerprint in fingerprints:
+            values: Sequence[str] = [fingerprint]
+
+            var = parse_fingerprint_var(fingerprint)
+            if var == "default":
+                values = self.handle_default_fingerprint(span)
+
+            result.extend(values)
+
+        return " ".join(result)
+
+    def handle_default_fingerprint(self, span: Span) -> Sequence[str]:
+        span_group = None
+
+        # Try using all of the strategies in order to generate
+        # the appropriate span group. The first strategy that
+        # successfully generates a span group will be chosen.
+        for strategy in self.strategies:
+            span_group = strategy(span)
+            if span_group is not None:
+                break
+
+        # If no strategies generated a valid span group,
+        # fall back to using the raw description strategy
+        if span_group is None:
+            span_group = raw_description_strategy(span)
+
+        return span_group
+
+
+def span_op(op_name: str) -> Callable[[CallableStrategy], CallableStrategy]:
+    def wrapped(fn: CallableStrategy) -> CallableStrategy:
+        return lambda span: fn(span) if span.get("op") == op_name else None
+
+    return wrapped
+
+
+def raw_description_strategy(span: Span) -> Sequence[str]:
+    """The catch-all strategy to use if all other strategies fail. This
+    strategy is only effective if the span description is a fixed string.
+    Otherwise, this strategy will produce a large number of span groups.
+    """
+    return [span.get("description") or ""]
```

Comment:
Where are these strings `md5` hashed? In Sentry or in Snuba?

Reply:
Right now, the intent is to hash these strings in the Snuba consumer, though I'm still evaluating where the best place for it is.

Reasons to do it in Snuba:
- hide the hash function as an implementation detail within snuba
- store the human-readable text to facilitate debugging

Reasons to do it in Sentry
- we can use the hash from the snuba results to perform a lookup in transaction from nodestore

Reply:
Also reduces the payload sent over the network if it's done in Sentry.

I have no strong feelings either way, I just didn't see where it was happening.

Reply:
I've opted to switch back to computing the hash in Sentry. This is primarily because we still need the hash to perform the lookup so we'll need the hash in nodestore.